### PR TITLE
Write_multi to use upsert_all

### LINKED ISSUE #40: write_multi operation writes cache to database with multiple inserts.
# Steps to repro:
1. Make sure, that Rails cache is configured:
`config.cache_store = ActiveSupport::Cache::DatabaseStore.new`
2. Write multiple cache entries:
`Rails.cache.write_multi({'test1' => 'test1', 'test2' => 'test2', 'test3' => 'test3'})`

# Result
```ruby
ActiveSupport::Cache::DatabaseStore::Model Load (1.3ms)  SELECT "activesupport_cache_entries".* FROM "activesupport_cache_entries" WHERE "activesupport_cache_entries"."key" = '\x7465737431' ORDER BY "activesupport_cache_entries"."key" ASC LIMIT 1 /*application:Cheddar*/
  TRANSACTION (0.4ms)  BEGIN /*application:Cheddar*/
  ActiveSupport::Cache::DatabaseStore::Model Create (0.8ms)  INSERT INTO "activesupport_cache_entries" ("key", "value", "version", "created_at", "expires_at") VALUES ('\x7465737431', '\x040849220a7465737431063a064554', NULL, '2023-09-14 10:55:21.040251', NULL) RETURNING "key" /*application:Cheddar*/
  TRANSACTION (0.4ms)  COMMIT /*application:Cheddar*/
  ActiveSupport::Cache::DatabaseStore::Model Load (0.4ms)  SELECT "activesupport_cache_entries".* FROM "activesupport_cache_entries" WHERE "activesupport_cache_entries"."key" = '\x7465737432' ORDER BY "activesupport_cache_entries"."key" ASC LIMIT 1 /*application:Cheddar*/
  TRANSACTION (0.4ms)  BEGIN /*application:Cheddar*/
  ActiveSupport::Cache::DatabaseStore::Model Create (0.6ms)  INSERT INTO "activesupport_cache_entries" ("key", "value", "version", "created_at", "expires_at") VALUES ('\x7465737432', '\x040849220a7465737432063a064554', NULL, '2023-09-14 10:55:21.044673', NULL) RETURNING "key" /*application:Cheddar*/
  TRANSACTION (0.5ms)  COMMIT /*application:Cheddar*/
  ActiveSupport::Cache::DatabaseStore::Model Load (0.7ms)  SELECT "activesupport_cache_entries".* FROM "activesupport_cache_entries" WHERE "activesupport_cache_entries"."key" = '\x7465737433' ORDER BY "activesupport_cache_entries"."key" ASC LIMIT 1 /*application:Cheddar*/
  TRANSACTION (0.4ms)  BEGIN /*application:Cheddar*/
  ActiveSupport::Cache::DatabaseStore::Model Create (0.4ms)  INSERT INTO "activesupport_cache_entries" ("key", "value", "version", "created_at", "expires_at") VALUES ('\x7465737433', '\x040849220a7465737433063a064554', NULL, '2023-09-14 10:55:21.053842', NULL) RETURNING "key" /*application:Cheddar*/
  TRANSACTION (0.3ms)  COMMIT /*application:Cheddar*/
=> {"test1"=>#<ActiveSupport::Cache::Entry:0x000000010b410dc8 @created_at=0.0, @expires_in=nil, @value="test1", @version=nil>,
 "test2"=>#<ActiveSupport::Cache::Entry:0x000000010b410d78 @created_at=0.0, @expires_in=nil, @value="test2", @version=nil>,
 "test3"=>#<ActiveSupport::Cache::Entry:0x000000010b410c88 @created_at=0.0, @expires_in=nil, @value="test3", @version=nil>}
```

# expected result
I would expect it to use upsert_all instead and do everything in one go.

### PR DESCRIPTION
**Pull request by skatkov**
fixes #40

### CODE REVIEW

File: lib/active_support/cache/database_store.rb [hunk 1]
```diff
@@ -120,6 +120,22 @@ def delete_entry(key, _options = nil)
         @model.where(key: key).destroy_all
       end
 
+      def write_multi_entries(hash, **_options)
+        entries = hash.each_with_object([]) do |(key, entry), obj|
```
Comment (dim) [inline lib/active_support/cache/database_store.rb]: you can just use `map`

File: spec/active_support/cache/database_store_spec.rb [hunk 2]
```diff
@@ -107,6 +107,21 @@
     expect { subject.read("#{key}x") }.to raise_error(ArgumentError, /exceeds the length limit/)
   end
 
+  it 'write_multi to insert records with one insert' do
+    insert_queries = []
+    ActiveSupport::Notifications.subscribe('sql.active_record') do |_name, _start, _finish, _id, payload|
+      insert_queries << payload[:sql] if payload[:sql].start_with?('INSERT INTO')
```
Comment (dim) [inline spec/active_support/cache/database_store_spec.rb:113]: I'm not sure it's the best way to check this. I would prefer just to do:

```rb
subject.write_multi(values)
expect(subject.read_multi('test0', 'test1', 'test2')).to eq(values)

expect(Model.where(key: values.keys).all.map(&:to_hash)).to match_array([
  ...
])
```

**Review by dim — CHANGES_REQUESTED** [review]

**Comment by dim** [comment]
Sorry, could you pls also update changelog

File: spec/active_support/cache/database_store_spec.rb (hunk 2 above)
Comment (skatkov) [reply spec/active_support/cache/database_store_spec.rb:113]: I'm just counting number of INSERT queries here, because my goal is to have only one :)

**Comment by skatkov** [comment]
@dim do you have idea what is from with sqlite on older Rails 6 branch?

File: spec/active_support/cache/database_store_spec.rb (hunk 2 above)
Comment (dim) [reply spec/active_support/cache/database_store_spec.rb:113]: sure, but I would also like to check that the written records are "complete", i.e. contain all the expected attributes, hence `match_array` which supports pattern matching

File: lib/active_support/cache/database_store.rb [hunk 3]
```diff
@@ -120,6 +120,23 @@ def delete_entry(key, _options = nil)
         @model.where(key: key).destroy_all
       end
 
+      def write_multi_entries(hash, **_options)
+        entries = hash.map do |key, entry|
+          expires_at = Time.zone.at(entry.expires_at) if entry.expires_at
+
+          {
```
Comment (dim) [inline lib/active_support/cache/database_store.rb]: maybe:

```rb
{
  key: key,
  created_at: Time.zone.now,
  **entry_attributes(entry)
}
```

Comment (dim) [reply lib/active_support/cache/database_store.rb]: then you can also change `write_entry` to:

```rb
record = @model.where(key: key).first_or_initialize
record.update!(**entry_attributes(entry))
```

File: spec/active_support/cache/database_store_spec.rb (hunk 2 above)
Comment (skatkov) [reply spec/active_support/cache/database_store_spec.rb:113]: I don't think this works the way you think this works.

Screenshot 2023-09-19 at 16 34 18

**Comment by skatkov** [comment]
@dim you want to take over from here? I can't figure out what's wrong with sqlite, it doesn't work with binary key properly in with rails 6.

File: lib/active_support/cache/database_store.rb [hunk 4]
```diff
@@ -126,17 +125,21 @@ def write_multi_entries(hash, **_options)
...
             created_at: Time.zone.now,
+            **entry_attributes(entry),
           }
         end
 
         # In rails 7, we can use update_only not to do anything. But for the sakes of compatibility, we don't use any additional parameters.
         @model.upsert_all(entries)
       end
 
+      def entry_attributes(entry)
```
Comment (dim) [inline lib/active_support/cache/database_store.rb:131]: should be private

Comment (skatkov) [reply lib/active_support/cache/database_store.rb:131]: It is, all these methods are listed under "private" directive.

**Comment by dim** [comment]
Thanks!

**Pull request merged**